osu.Game/Screens/Play/SubmittingPlayer.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Threading.Tasks;
+using osu.Framework.Allocation;
+using osu.Framework.Logging;
+using osu.Framework.Screens;
+using osu.Game.Online.API;
+using osu.Game.Online.Rooms;
+using osu.Game.Scoring;
+
+namespace osu.Game.Screens.Play
+{
+    /// <summary>
+    /// A player instance which supports submitting scores to an online store.
+    /// </summary>
+    public abstract class SubmittingPlayer : Player
+    {
+        protected long? Token { get; private set; }
+
+        [Resolved]
+        private IAPIProvider api { get; set; }
+
+        protected SubmittingPlayer(PlayerConfiguration configuration = null)
+            : base(configuration)
+        {
+        }
+
+        protected override void LoadAsyncComplete()
+        {
+            // Token request construction should happen post-load to allow derived classes to potentially prepare DI backings that are used to create the request.
+            var tcs = new TaskCompletionSource<bool>();
+
+            if (!api.IsLoggedIn)
+            {
+                fail(new InvalidOperationException("API is not online."));
+                return;
+            }
+
+            var req = CreateTokenRequestRequest();
+
+            if (req == null)
+            {
+                fail(new InvalidOperationException("Request could not be constructed."));
+                return;
+            }
+
+            req.Success += r =>
+            {
+                Token = r.ID;
+                tcs.SetResult(true);
+            };
+            req.Failure += fail;
+
+            api.Queue(req);
+
+            tcs.Task.Wait();
```

Comment:
Slightly worried about this wait. Doesn't seem to deadlock on desktop, as this wait appears to be cross-thread even in single-threaded mode:

* `LoadCompleteAsync()` is on a background load thread
* API request callbacks are on the main thread

but I'd probably run this past something with a synchronisation context (so android, at this point) to be sure. May still deadlock if main thread higher up is waiting for this to complete, which doesn't sound impossible.

Reply:
This hasn't changed from the previous implementation, so it *should* be fine. That said, if a `Player` is ever loaded synchronously in a case where we are waiting on this API request, there is definitely a potential of deadlock (the `Success` invocation gets sceduled to `APIAccess`'s scheduler, which is running on the update thread.

Not sure if it should be of immediate concern. It's hard to resolve while keeping the logic intentions of this flow (to delay the actual load process until the request is complete).

Reply:
If you mean the previous implementation in `PlaylistsPlayer`, then that *has* changed in a meaningful way (went from polling a field in a loop, to blocking on an async task). The older method was uglier in general, but "safer" from a threading perspective, as there was no synchronisation context-related deadlock risk.

Reply:
The `SynchronisationContext` specific case should already be handled via `ConfigureAwait` being used everywhere, I think.